src/components/course/CoursePage.jsx

```diff
@@ -1,12 +1,35 @@
-import React from 'react';
+import React, { useContext, useEffect } from 'react';
+import Cookies from 'universal-cookie';
+import qs from 'query-string';
+import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
+import { getConfig } from '@edx/frontend-platform/config';
+import { AppContext } from '@edx/frontend-platform/react';
 
 import Course from './Course';
 import AuthenticatedUserSubsidyPage from '../app/AuthenticatedUserSubsidyPage';
+import { loginRefresh } from '../../utils/common';
 
 export default function CoursePage() {
-  return (
-    <AuthenticatedUserSubsidyPage>
-      <Course />
-    </AuthenticatedUserSubsidyPage>
-  );
+  const { authenticatedUser } = useContext(AppContext);
+  const { roles } = authenticatedUser;
+
+  // If the user has not refreshed their JWT since they created their account,
+  // we should refresh it so that they'll have appropriate roles (if available),
+  // and thus, have any appropriate permissions when making downstream requests.
+  if (!roles || Object.keys(roles).length == 0) {
+    loginRefresh().then(() => {
```

Comment:
`loginRefresh` isn't currently returning anything. I think you'll need to return the await'ed promise if you want to make `loginRefresh` then-able. You might want to introduce a state variable that gets set when `loginRefresh` is fully resolved, and only render the `AuthenticatedUserSubsidyPage` and `Course` components when that state variable passes a condition. You also don't want to make this call on every single re-render of `CoursePage`, only when `roles` change.

Maybe something along the lines of the following? While the login refresh is pending, it displays a loading spinner in case of a slow network response. The `loginRefresh` function returns the awaited promise, or in the `catch`, returns a promise that immediately resolves. The `loginRefresh` is called in a `useEffect` to ensure it only gets called once (or only when `isRefreshingJWT` changes and is true). It also uses the shorthand of getting the `authenticatedUser` out of the context via the `getAuthenticatedUser` function.

```jsx
export const loginRefresh = async () => {
  const config = getConfig();
  const loginRefreshUrl = `${config.LMS_BASE_URL}/login_refresh`;
  try {
    return await getAuthenticatedHttpClient().post(loginRefreshUrl);
  } catch (error) {
    const isUserUnauthenticated = error.response?.status === 401;
    if (isUserUnauthenticated) {
      // Clean up the cookie if it exists to eliminate any situation
      // where the cookie is not expired but the jwt is expired.
      const cookies = new Cookies();
      cookies.remove(config.ACCESS_TOKEN_COOKIE_NAME);
    }
    return Promise.resolve();
  }
};
```

```jsx
// additional necessary imports:
import { getAuthenticatedUser } from '@edx/frontend-platform/auth';
import { Container } from '@edx/paragon';
import { LoadingSpinner } from '../loading-spinner';

export default function CoursePage() {
  const authenticatedUser = getAuthenticatedUser();
  const { roles } = authenticatedUser;
  // If the user has not refreshed their JWT since they created their account,
  // we should refresh it so that they'll have appropriate roles (if available),
  // and thus, have any appropriate permissions when making downstream requests.
  const [isRefreshingJWT, setIsRefreshingJWT] = useState(roles.length === 0);

  useEffect(() => {
    const refreshJWT = async () => {
      await loginRefresh();
      setIsRefreshingJWT(false);
    };
    if (isRefreshingJWT) {
      refreshJWT();
    }
  }, [isRefreshingJWT]);

  if (isRefreshingJWT) {
    return (
      <Container size="lg" className="py-5">
        <LoadingSpinner screenReaderText="loading user details" />
      </Container>
    );
  }

  return (
    <AuthenticatedUserSubsidyPage>
      <Course />
    </AuthenticatedUserSubsidyPage>
  );
}
```